libbeat/registry/backend/memlog/memlog.go

```diff
@@ -0,0 +1,176 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memlog
+
+import (
+	"math"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/registry/backend"
+)
+
+// Registry provides access to the memlog stores.
+// Each store will have it's own directory with data files and transaction log.
+type Registry struct {
+	mu     sync.Mutex
+	active bool
+
+	// root     string
+	// fileMode os.FileMode
+	settings Settings
+
+	wg sync.WaitGroup
+}
+
+// Settings configures a new Registry.
+type Settings struct {
+	// Registry root directory. Stores will be single sub-directories.
+	Root string
+
+	// FileMode is used to configure the file mode for new files generated by the regisry.
+	// File mode 0600 will be used if this field is not set.
+	FileMode os.FileMode
+
+	// BufferSize configures the IO buffer size when accessing the underlying storage files.
+	// Defaults to 4096 if not set.
+	BufferSize uint
+
+	// Checkpoint predicates is used to check when the registry file should run a checkpoint operation.
+	// On checkpoint a new snapshot of the current in memory state is written to
+	// a new data file and the transaction log is reset.
+	// Checkpoint must return true, in order to trigger a new snapshot.
+	// Defaults to CheckpointRatio(2, 10).
+	Checkpoint func(pairs, logs uint) bool
+}
+
+const defaultFileMode os.FileMode = 0600
+const defaultBufferSize = 4096
+
+var defaultCheckpoint = CheckpointRatio(2.0, 10.0)
+
+// New creates a new Registry.
+func New(settings Settings) (*Registry, error) {
+	if settings.FileMode == 0 {
+		settings.FileMode = defaultFileMode
+	}
+	if settings.Checkpoint == nil {
+		settings.Checkpoint = defaultCheckpoint
+	}
+	if settings.BufferSize == 0 {
+		settings.BufferSize = defaultBufferSize
+	}
+
+	return &Registry{
+		active:   true,
+		settings: settings,
+	}, nil
+}
+
+// Access creates or opens a new store. A new sub-directory for the store if
+// created, if the store does not exist.
+// Returns an error is any file access fails.
+func (r *Registry) Access(name string) (backend.Store, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if !r.active {
+		return nil, errRegClosed
+	}
+
+	root := r.settings.Root
+	fileMode := r.settings.FileMode
+	bufSz := int(r.settings.BufferSize)
```

Comment:
Why is `BufferSize` converted here? I think is can be done later when passing the variable to `newDiskStore`. `diskStore` is the first type which requires the size to be an `int` to be able to call `bufio.NewWriterSize`.

Reply:
Doesn't really matter why it's converted.